File: Sources/PDFSigner/LicenseManager.swift
import Foundation
import Security

/// Minimal Keychain generic-password store for the license token. 🔒
enum LicenseKeychain {
    private static let service = "com.palbin.pdfsigner.license"
    private static let account = "license-key"

    static func read() -> String? {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(q as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func write(_ key: String) {
        let data = Data(key.utf8)
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = data
        SecItemAdd(add as CFDictionary, nil)
    }

    static func clear() {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(base as CFDictionary)
    }
}

/// License + free-tier metering. Pro is unlocked by a valid offline license
/// token (stored in the Keychain). The token issuance/Stripe flow lives on the
/// server side (not bundled); `validate(_:)` is the single seam to plug it in.
@MainActor
@Observable
final class LicenseManager {
    enum Tier: String { case free, pro }

    private(set) var tier: Tier = .free
    private(set) var monthlyCount: Int = 0

    /// Free tier: number of signatures per calendar month.
    let freeMonthlyLimit = 10

    private let defaults = UserDefaults.standard

    init() {
        if let key = LicenseKeychain.read(), Self.validate(key) {
            tier = .pro
        }
        monthlyCount = defaults.integer(forKey: countKey())
    }

    var isPro: Bool { tier == .pro }
    var remainingFreeSigns: Int { max(0, freeMonthlyLimit - monthlyCount) }

    /// Offline token check. A real deployment verifies a signed token against an
    /// embedded public key + expiry/grace; here we accept the documented format
    /// `PDFS-XXXX-XXXX-XXXX` so the gating/UI is exercised end-to-end.
    static func validate(_ key: String) -> Bool {
        let parts = key.uppercased().split(separator: "-")
        return parts.count == 4 && parts[0] == "PDFS" && parts.dropFirst().allSatisfy { $0.count == 4 }
    }

    @discardableResult
    func activate(_ key: String) -> Bool {
        guard Self.validate(key) else { return false }
        LicenseKeychain.write(key)
        tier = .pro
        return true
    }

    func deactivate() {
        LicenseKeychain.clear()
        tier = .free
    }

    /// Record one successful signature against the monthly free quota.
    func recordSign() {
        guard !isPro else { return }
        monthlyCount += 1
        defaults.set(monthlyCount, forKey: countKey())
    }

    private func countKey() -> String {
        let c = Calendar.current.dateComponents([.year, .month], from: Date())
        return "signCount-\(c.year ?? 0)-\(c.month ?? 0)"
    }
}

File: windows/BureaucratPdf.App/LicenseManager.cs
using System.Security.Cryptography;
using System.Text;
using Windows.Storage;

namespace BureaucratPdf.App;

/// <summary>
/// License + free-tier metering for Windows. Ports macOS LicenseManager, swapping
/// the Keychain for DPAPI (ProtectedData, CurrentUser) to store the token and
/// ApplicationData.LocalSettings for the monthly counter.
///
/// Pro is unlocked by a valid offline license token. The token issuance/Stripe
/// flow lives server-side (not bundled); Validate is the single seam to plug it in.
/// </summary>
public sealed class LicenseManager
{
    public enum Tier { Free, Pro }

    public Tier CurrentTier { get; private set; } = Tier.Free;
    public int MonthlyCount { get; private set; }

    /// <summary>Free tier: signatures per calendar month.</summary>
    public const int FreeMonthlyLimit = 10;

    public bool IsPro => CurrentTier == Tier.Pro;
    public int RemainingFreeSigns => Math.Max(0, FreeMonthlyLimit - MonthlyCount);

    private const string TokenSetting = "license.token";       // DPAPI-protected, base64
    private static ApplicationDataContainer Local => ApplicationData.Current.LocalSettings;

    public LicenseManager()
    {
        var token = ReadToken();
        if (token is not null && Validate(token)) CurrentTier = Tier.Pro;
        MonthlyCount = Local.Values[CountKey()] is int n ? n : 0;
    }

    /// <summary>
    /// Offline token check. A real deployment verifies a signed token against an
    /// embedded public key + expiry/grace; here we accept the documented format
    /// PDFS-XXXX-XXXX-XXXX so the gating/UI is exercised end-to-end.
    /// </summary>
    public static bool Validate(string key)
    {
        var parts = key.ToUpperInvariant().Split('-');
        return parts.Length == 4 && parts[0] == "PDFS" && parts.Skip(1).All(p => p.Length == 4);
    }

    public bool Activate(string key)
    {
        if (!Validate(key)) return false;
        WriteToken(key);
        CurrentTier = Tier.Pro;
        return true;
    }

    public void Deactivate()
    {
        Local.Values.Remove(TokenSetting);
        CurrentTier = Tier.Free;
    }

    /// <summary>Record one successful signature against the monthly free quota.</summary>
    public void RecordSign()
    {
        if (IsPro) return;
        MonthlyCount++;
        Local.Values[CountKey()] = MonthlyCount;
    }

    // ── DPAPI token storage (CurrentUser scope) ──────────────────────────────

    private static void WriteToken(string key)
    {
        var plain = Encoding.UTF8.GetBytes(key);
        var enc = ProtectedData.Protect(plain, optionalEntropy: null, DataProtectionScope.CurrentUser);
        Local.Values[TokenSetting] = Convert.ToBase64String(enc);
    }

    private static string? ReadToken()
    {
        if (Local.Values[TokenSetting] is not string b64) return null;
        try
        {
            var enc = Convert.FromBase64String(b64);
            var plain = ProtectedData.Unprotect(enc, optionalEntropy: null, DataProtectionScope.CurrentUser);
            return Encoding.UTF8.GetString(plain);
        }
        catch
        {
            return null; // tampered / different user — treat as no license
        }
    }

    private static string CountKey()
    {
        var now = DateTime.Now;
        return $"signCount-{now.Year}-{now.Month}";
    }
}
